test/test_propagate_convolution.py

```diff
@@ -0,0 +1,151 @@
+"""Test PyDynamic.uncertainty.propagate_convolve"""
+
+import numpy as np
+import pytest
+import scipy.ndimage as sn
+from hypothesis import assume, given, strategies as hst
+
+from PyDynamic.uncertainty.propagate_convolution import convolve_unc
+
+
+def random_array(length):
+    array = np.random.randn(length)
+    return array
+
+
+def random_covariance_matrix(length):
+    """construct a valid (but random) covariance matrix with good condition number"""
+
+    # because np.cov estimates the mean from data, the returned covariance matrix
+    # has one eigenvalue close to numerical zero (rank n-1).
+    # This leads to a singular matrix, which is badly suited to be used as valid
+    # covariance matrix. To circumvent this:
+
+    ## draw random (n+1, n+1) matrix
+    cov = np.cov(np.random.random((length + 1, length + 1)))
+
+    ## calculate SVD
+    u, s, vh = np.linalg.svd(cov, full_matrices=False, hermitian=True)
+
+    ## reassemble a covariance of size (n, n) by discarding the smallest singular value
+    cov_adjusted = (u[:-1, :-1] * s[:-1]) @ vh[:-1, :-1]
+
+    return cov_adjusted
+
+
+def valid_inputs(reduced_set=False):
+
+    valid_inputs = []
+
+    for n in [10, 15, 20]:
+        x_signal = random_array(n)
+        u_signal = random_covariance_matrix(n)
+
+        if reduced_set:
+            valid_inputs.append([x_signal, u_signal])
+        else:
+            valid_inputs.append([x_signal, None])
+            valid_inputs.append([x_signal, np.diag(np.diag(u_signal))])
+            valid_inputs.append([x_signal, u_signal])
+
+    return valid_inputs
+
+
+def valid_modes(kind="all"):
+    scipy_modes = ["nearest", "reflect", "mirror"]
+    numpy_modes = ["full", "valid", "same"]
+
+    if kind == "all":
+        return numpy_modes + scipy_modes
+    elif kind == "scipy":
+        return scipy_modes
+    elif kind == "numpy":
+        return numpy_modes
+    else:
+        return []
+
+
+@pytest.mark.parametrize("input_1", valid_inputs())
+@pytest.mark.parametrize("input_2", valid_inputs())
+@pytest.mark.parametrize("mode", valid_modes())
+def test_convolution(input_1, input_2, mode):
+
+    scipy_modes = valid_modes("scipy")
+    numpy_modes = valid_modes("numpy")
+
+    # calculate the convolution of x1 and x2
+    y, Uy = convolve_unc(*input_1, *input_2, mode)
+
+    if mode in numpy_modes:
+        y_ref = np.convolve(input_1[0], input_2[0], mode=mode)
+    elif mode in scipy_modes:
+        y_ref = sn.convolve1d(input_1[0], input_2[0], mode=mode)
+    else:
+        raise ValueError(f"Unsupported Mode <{mode}>")
```

Comment:
I somehow do not like the idea of assuming the test case is wrong and throwing an exception in a test case itself. One basic assumption is the test must be correct. This else-branch relies on quite some assumptions, mainly we know that any valid mode needs to be part of either what we expect to get from `valid_modes("scipy")` and `valid_modes("numpy")`. I would rather strip this logic from the test. Anyway this is what probably happens when calling Numpy or Scipy with a wrong mode, isn't it?

Reply:
Good point, I removed the `else`-branch.